seed/validate.go

```diff
@@ -24,62 +24,115 @@ import (
 	"fmt"
 	"path/filepath"
 	"regexp"
+	"sort"
 
-	"github.com/snapcore/snapd/seed/internal"
 	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/timings"
 )
 
+type ValidationError struct {
+	// SystemErrors maps system labels ("" for UC16/18) to their validation errors.
+	SystemErrors map[string][]error
+}
+
+func newValidationError(label string, err error) *ValidationError {
+	return &ValidationError{SystemErrors: map[string][]error{
+		label: {err},
+	}}
+}
+
+func (e *ValidationError) addErr(label string, errs ...error) {
+	if e.SystemErrors == nil {
+		e.SystemErrors = make(map[string][]error)
+	}
+	for _, err := range errs {
+		e.SystemErrors[label] = append(e.SystemErrors[label], err)
+	}
+}
+
+func (e ValidationError) hasErrors() bool {
+	return len(e.SystemErrors) != 0
+}
+
+func (e *ValidationError) Error() string {
+	systems := make([]string, 0, len(e.SystemErrors))
+	for s := range e.SystemErrors {
+		systems = append(systems, s)
+	}
+	sort.Strings(systems)
+	var buf bytes.Buffer
+	first := true
+	for _, s := range systems {
+		if first {
+			if s == "" {
+				fmt.Fprintf(&buf, "cannot validate seed:")
+			} else {
+				fmt.Fprintf(&buf, "cannot validate seed system %q:", s)
```

Comment:
is this else clause tested, or is it UC20 followup material?

Reply:
it's follow up material, I actually wanted to do one PR but it seems we want the UC 16/18 bits sooner if possible.  We basically cannot reach this code with current usage. I could undo it but not sure it's material.